waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,195 @@
+            [saml20-clj.shared :as saml-shared]
+            [saml20-clj.sp :as saml-sp]
+            [taoensso.nippy :as nippy]
+            [waiter.auth.authentication :as auth]
+            [waiter.middleware :as middleware]
+            [waiter.util.utils :as utils]))
+
+(defrecord SamlAuthenticator [auth-redirect-uri idp-cert idp-uri password saml-acs-handler-fn saml-auth-redirect-handler-fn saml-req-factory!]
+  auth/Authenticator
+  (wrap-auth-handler [_ request-handler]
+    (fn saml-authenticator-handler [{:keys [headers query-string request-method uri] :as request}]
+      (let [waiter-cookie (auth/get-auth-cookie-value (get headers "cookie"))
+            [auth-principal _ :as decoded-auth-cookie] (auth/decode-auth-cookie waiter-cookie password)]
+        (cond
+          ;; Use the cookie, if not expired
+          (auth/decoded-auth-valid? decoded-auth-cookie)
+          (let [auth-params-map (auth/auth-params-map auth-principal)
+                request-handler' (middleware/wrap-merge request-handler auth-params-map)]
+            (request-handler' request))
+          :else
+          (case request-method
+            :get (let [saml-request (saml-req-factory!)
+                       relay-state (str (name (utils/request->scheme request)) "://" (get headers "host") uri (if query-string "?" "") query-string)]
+                   (saml-sp/get-idp-redirect idp-uri saml-request relay-state))
+            :post (if-let [saml-auth-data (get-in (ring-params/params-request request) [:form-params "saml-auth-data"])]
+                    (let [{:keys [not-on-or-after saml-principal]}
+                          (try
+                            (nippy/thaw (b64/decode (.getBytes saml-auth-data))
+                                        {:password password :v1-compatibility? false :compressor nil})
+                            (catch Exception e
+                              (throw (ex-info "Could not parse saml-auth-data." {:status 400
+                                                                                 :saml-auth-data saml-auth-data
+                                                                                 :inner-exception e}))))
+                          t-now (t/now)
+                          _ (when-not (t/before? t-now not-on-or-after)
+                              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                                              {:status 400
+                                               :saml-assertion-not-on-or-after not-on-or-after
+                                               :t-now t-now})))
+                          age-in-seconds (t/in-seconds (t/interval t-now (t/min-date (t/plus t-now (t/days 1)) not-on-or-after)))
+                          {:keys [authorization/principal authorization/user] :as auth-params-map}
+                          (auth/auth-params-map saml-principal)
+                          request' (-> request
```

Comment:
we discussed triggering a redirect from here.

Reply:
done